### DISCUSSION TITLE
Parallel Routes not working on Cloud Run in Next.js 15

### DISCUSSION THREAD

**Discussion by aramikuto** [vercel/next.js#73539] [Help]
### Link to the code that reproduces this issue

https://github.com/aramikuto/nextjs-parallel-route-server-js-problem

### To Reproduce

1. Build the application and start the standalone server with the following command:
```bash
yarn build && cp -r .next/static .next/standalone/.next/ && node .next/standalone/server.js
```
2. Open the network inspector and navigate to the homepage at http://localhost:3000/ 
3. Note the URL of the chunk containing the parallel route.

### Current vs. Expected behavior

In Next.js 14, the @ symbol in the chunk URL is not encoded, appearing as:
http://localhost:3000/_next/static/chunks/app/@modal/page-5d9f089824f6dc9b.js

In Next.js 15.0.0 and 15.0.1-canary.1, the @ symbol is encoded as %40, so the URL looks like:
http://localhost:3000/_next/static/chunks/app/%40modal/page-e0360b9b769efb1a.js

When deployed to Cloud Run, this chunk becomes inaccessible because [Some encoded URL characters are decoded](https://cloud.google.com/run/docs/known-issues#http2-iap) by Cloud Run. It seems Cloud Run decodes %40 back to @, but the standalone server on Next.js 15 responds with a 404 error for such requests.

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.6.0: Mon Jul 29 21:16:46 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T8112
  Available memory (MB): 24576
  Available CPU cores: 8
Binaries:
  Node: 20.11.1
  npm: 10.2.4
  Yarn: 1.22.21
  pnpm: N/A
Relevant Packages:
  next: 15.0.1-canary.1 // Latest available version is detected (15.0.1-canary.1).
  eslint-config-next: N/A
  react: 19.0.0-rc-69d4b800-20241021
  react-dom: 19.0.0-rc-69d4b800-20241021
  typescript: 5.3.3
Next.js Config:
  output: standalone
```

### Which area(s) are affected? (Select all that apply)

Parallel & Intercepting Routes

### Which stage(s) are affected? (Select all that apply)

Other (Deployed)

### Additional context

It seems this issue may have been introduced in PR #70256.

**Comment by nicholas5538** [comment]
~Same issue here, the parallel route for opening my modal works only on development, but it does not work when I use `next start` or preview it on the Netlify deploy preview.~

~Code has been left unchanged after the update other than adding the `await` on my `params` prop~

My issue has been fixed by removing `export const dynamic = 'error';`, the solution was to make my parallel routes server-rendered on demand

**Comment by jeqermigselv-cloud** [comment]
I have the same issue in Azure. Works fine on my local machine, but when deployed to Azure, the files it not found (despite copying it)

**Comment by Alexandredc** [comment]
@aramikuto did you find any workaround ?

**Comment by mozisan** [comment]
@Alexandredc 
Hi, I found a workaround that resolved the issue in my environment.
Hope this helps you.

```ts
// middleware.ts

import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";

export const config = {
  matcher: "/_next/static/chunks/:path*",
};

export const middleware = (request: NextRequest) => {
  const url = new URL(request.url);

  const decodedPathname = url.pathname.split("/").map(decodeURIComponent).join("/");
  const encodedPathname = decodedPathname.split("/").map(encodeURIComponent).join("/");

  const includesEncodableChar = url.pathname !== encodedPathname;
  if (!includesEncodableChar) return;

  const hasAlreadyDecoded = url.pathname === decodedPathname;
  if (hasAlreadyDecoded) return;

  const destination = new URL(url);
  destination.pathname = encodedPathname;
  return NextResponse.rewrite(destination);
};
```

**Reply by omerman** [reply]
Thank you so much! I can confirm this fix works with cloudfront. (I had the exact issue there)

**Reply by ipravdin-opog** [reply]
Also it worked for our set up. For others to note, as it was an issue for me, if you have localisation in the app with dynamic route `[locale]`, you probably ignore `"/_next/static/"` in the matcher

**Comment by Alexandredc** [comment]
> @Alexandredc Hi, I found a workaround that resolved the issue in my environment. Hope this helps you.
> 
> ```ts
> // middleware.ts
> 
> import { NextResponse } from "next/server";
> import type { NextRequest } from "next/server";
> 
> export const config = {
>   matcher: "/_next/static/chunks/:path*",
> };
> 
> export const middleware = (request: NextRequest) => {
>   const url = new URL(request.url);
> 
>   const decodedPathname = url.pathname.split("/").map(decodeURIComponent).join("/");
>   const encodedPathname = decodedPathname.split("/").map(encodeURIComponent).join("/");
> 
>   const includesEncodableChar = url.pathname !== encodedPathname;
>   if (!includesEncodableChar) return;
> 
>   const hasAlreadyDecoded = url.pathname === decodedPathname;
>   if (hasAlreadyDecoded) return;
> 
>   const destination = new URL(url);
>   destination.pathname = encodedPathname;
>   return NextResponse.rewrite(destination);
> };
> ```

I have tried in Google Cloud Run and it doesn't work :(

**Comment by Alexandredc** [comment]
@ForsakenHarmony Do you have some guidance to implement a workaround, please? 🙏

**Comment by aramikuto** [comment]
@Alexandredc I'm currently waiting on a response from Vercel to confirm if they recognize this as an issue. If you need to upgrade right away, it might be worth exploring options for deploying static assets separately; for example, you could use Firebase Hosting to serve static resources. I haven’t checked yet, but hopefully, Firebase Hosting isn’t impacted by the issue.

**Comment by rollsrobby** [comment]
We have the same running in a Docker container in Azure. It is currently blocking us from upgrading. Unfortunately the middleware workaround doesn't work for us.

**Comment by Alexandredc** [comment]
@aramikuto do you have any news from Vercel ?
I tried to find a workaround today, but was unsuccessful. :(

**Comment by Alexandredc** [comment]
Finally, I have found a simple solution that work on Google Cloud Run.

Add a rewrite rule on `next.config.ts`

```ts
import type { NextConfig } from 'next';

const nextConfig: NextConfig = {
  output: 'standalone',
  rewrites: async () => {
    return {
      beforeFiles: [
        {
          source: '/_next/static/chunks/app/:folder/@breadcrumb/:path*',
          destination: '/_next/static/chunks/app/:folder/%40breadcrumb/:path*'
        }
      ],
      afterFiles: [],
      fallback: []
    };
  }
};

export default nextConfig;
```

**Comment by aramikuto** [comment]
@Alexandredc unfortunately no news yet. Thank you for sharing the workaround!

**Comment by ru-van-urk** [comment]
For anyone also experiencing this on Google Cloud Run

For me @Alexandredc workaround didn't work because I was using both dynamic routes and route groups. My path looks like this: `/_next/static/chunks/app/%5Blocale%5D/admin/(routes)/%40employer/layout-fcdbade18a260943.js`


I fixed it by adding a wildcard to the folder slug:

```js
import createNextIntlPlugin from "next-intl/plugin";

const withNextIntl = createNextIntlPlugin();

/** @type {import("next").NextConfig} */
const config = {
  output: "standalone",
  rewrites: async () => ({
    beforeFiles: [
      {
       // Added wildcard to ':folder' -> ':folder*'
        source: "/_next/static/chunks/app/:folder*/@employer/:path*",
        destination: "/_next/static/chunks/app/:folder*/%40employer/:path*",
      },
      // more parrallel rewrites
    ],
    afterFiles: [],
    fallback: [],
  }),
};

export default withNextIntl(config);
```

**Comment by samcx** [comment]
Hi everyone, was investigating this—

The [PR](https://github.com/vercel/next.js/pull/70256) that is causing this issue for Azure and Cloud Run (as far as I'm aware) is because both platforms are (I don't see why they should be, I would consider that a bug/unusual to do) decoding the request path in-flight. The Next.js PR fixed an issue where the pathing was not consistent (described in the PR description).

- https://github.com/Azure/azure-functions-host/pull/9402
- https://cloud.google.com/run/docs/known-issues#url-decode

Let me know if you have any further questions!

**Reply by macrozone** [reply]
Is there any plan to fix this in nextjs with a workaround? I understand that its not nextjs' fault here, but cloud run and azure are too big to ignore.

next 15 basically removed support for those platforms, I think this is too big of an issue to just convert it to a discussion. I suggest to reopen the issue and find ways to fix this.

**Comment by masoudmanson** [comment]
Hi there! Any updates on this issue yet? I’m still experiencing the same problem with Next.js 15.0.3 on Azure.

Thanks in advance!

**Comment by datoon83** [comment]
Is there any further progress on this?

We've double encoded for now, but seems like a recipe for breaking further down the line.

**Comment by anarqz** [comment]
Bumping the discussion since I've moved my entire app to parallel routes just to discover that it doesn't work in GCP 😮‍💨

**Comment by maral** [comment]
Another bump, upgraded to Next 15 only to find out parallel routes are broken... 🙈

**Reply by aramikuto** [reply]
Can you confirm that the problem persists with the latest versions of the framework (15.2.x or 15.3.x)? In the latest versions, the problem mentioned in this issue no longer occurs for me.

**Reply by SarahVanDenBerghe** [reply]
I'm using Next 15.3.0, having the same issue.

**Reply by maral** [reply]
In my case I've got the latest 15.3.1, and it happens on a managed server (not Azure/GCP).

**Comment by SarahVanDenBerghe** [comment]
I tried the solutions suggested above, but they didn’t resolve the issue in our case. In our case, the issue was caused by the proxy decoding parts of the URL, this led to broken requests on the Next.js server. Our backend partner resolved it by configuring the proxy to forward the original, unmodified URL. For our setup (Apache), this was done by adding the `nocanon` option to the proxy configuration.

**Reply by Shadowbring** [reply]
That did the trick for us, thanks! ❤️

**Comment by macrozone** [comment]
just run into this with google cloud run and was very surprised that such a basic thing like serving static files is such an issue.  :-/

**Comment by davemecha** [comment]
I experienced a similar (maybe the same) issue. In my case it was caused by an additional Next.js middleware, that was added by `next-intl`. There is also an issue (https://github.com/amannn/next-intl/issues/255) in the `next-intl` project for this. But I could imagine, a similar issue could happen with other middlewares as well. 

In my case my app was crashing due to client rendered parallel routes failing in my production environment with a ChunkLoadError. I deployed my app on Firebase App Hosting, which is based on Google Cloud Run.

The comments https://github.com/vercel/next.js/discussions/73539#discussioncomment-11467377 and https://github.com/vercel/next.js/discussions/73539#discussioncomment-11467379 inspired me to digg deeper and create a demo project (https://github.com/davemecha/nextjs-parallel-routes-crash-mcve) describing and reproducing the issue, with a generalized workaround.

My workaround is similar to the ones in the mentioned comments from above, but I used a variable to catch all possible slot names. So this rewrite rule should (hopefully work) for any slot name.

``` ts 
const nextConfig: NextConfig = {
  rewrites: async () => ({
    beforeFiles: [
      {
        source: '/_next/static/chunks/app/:folder*/@:slotName/:path*',
        destination: '/_next/static/chunks/app/:folder*/%40:slotName/:path*',
      },
    ],
    afterFiles: [],
    fallback: [],
  }),
};

const withNextIntl = createNextIntlPlugin();
export default withNextIntl(nextConfig);
```

I'm still not sure if this is really the same issue or it just seems to be. But maybe the fix helps someone.

**Comment by spasem** [comment]
I just hit this problem on GCP. I use Firestore Hosting with Cloud Run and all of a sudden I started getting the 404 on route paths with special characters. In my case it was static files in dynamic routes containing parentheses in the URL. I ended up adding the middleware file below.  My first thought was because of blitz, hence the header change as well. Not sure it's necessary. You can also add any special characters with currentPath.includes().

```
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  const currentPath = request.nextUrl.pathname

  // TARGET: Static files that look like they have encoded parentheses
  // We check for %28 OR if the path starts with _next/static
  if (
    currentPath.startsWith('/_next/static') || 
    currentPath.includes('%28')
  ) {
    
    // 1. DECODE THE PATH
    // Change /_next/.../%28main%29/... -> /_next/.../(main)/...
    // This fixes the "File Not Found" error.
    const decodedPath = decodeURIComponent(currentPath)
    
    // 2. SPOOF THE HOST
    // This fixes the "Blitz Security" error.
    const requestHeaders = new Headers(request.headers)
    requestHeaders.set('x-forwarded-host', 'localhost')
    requestHeaders.set('host', 'localhost')

    // 3. REWRITE (Internal Redirect)
    // We rewrite the request to the new DECODED path with the new SAFE headers.
    return NextResponse.rewrite(new URL(decodedPath, request.url), {
      request: {
        headers: requestHeaders,
      },
    })
  }

  return NextResponse.next()
}

// FORCE RUN: Run on every request to catch the encoded paths
export const config = {
  matcher: '/:path*',
}
```
